File: packages/pf_io/lib/src/export_payload.dart
/// 导出载荷（NDJSON）编码器 —— 规格 §3.2 / §4.2 3.3–3.5。
///
/// 职责单一：把「调用方给出的各阶段记录」编成**字节确定**的 NDJSON 载荷。
/// 字节确定不是审美问题 —— manifest 的 `contentHash` 覆盖全部非 manifest 行，
/// 行序或序列化的任何漂移都会让它失配；导入端拿它做载荷完整性裁决。
///
/// 三条硬规则（全部被 `export_payload` 向量锁死，期望值由
/// `tools/golden_vectors_gen/export_payload.py` 独立实现生成）：
///
///   1. **行序 = manifest → 按 [kPayloadStageOrder] 的记录行 → end**。
///      父实体先行（ledger → account → category → tag → theme → txn →
///      budget → attachment），保证导入时引用已存在。
///   2. **contentHash = SHA256(全部记录行的 UTF-8 字节，按出现顺序)**。
///      规格原文写「覆盖所有非 manifest 行」，但 end 行自身携带同一哈希 ——
///      若把 end 也算进覆盖，就是自指。因此口径固定为「记录行」
///      （2026-09-17 裁决）：manifest 与 end 均不在覆盖内；
///      end 行携带同一哈希与 recordCount，供导入端流式解析到最后一行时
///      与实际累计值比对。
///   3. **manifest 的 contentHash 由本编码器注入**。调用方给的 manifest 若已
///      含该键一律拒绝 —— 允许传入等于允许伪造。
///   4. **判别键 `type` 在记录行里独占**（2026-09-21 裁决，见
///      [kPayloadDiscriminatorKey]）。业务列在载荷层改名：
///      `account.type → accountType`、`txn.type → txnType`。
///      记录字段携带 `type` 一律 `PFC_E_VALIDATION` 拒绝。
library;

import 'dart:convert';
import 'dart:typed_data';

import 'package:pf_core/pf_core.dart';
import 'package:pf_crypto/pf_crypto.dart';

/// 记录行的输出阶段与顺序（§4.2 3.4 的 `kExportStageOrder`）。
///
/// 行序是**载荷格式契约**的一部分：导入端按行流式解析、逐条 upsert，
/// 乱序载荷会让「txn 引用尚未导入的 account」这类引用悬空。
/// 行内 JSON 对象的键序不在契约内（C1 规则：未知字段/键序差异必须容忍），
/// 但编码器输出是确定的（Map 插入序），这只是为了字节级可比，不是承诺。
const List<String> kPayloadStageOrder = <String>[
  'ledger',
  'account',
  'category',
  'tag',
  'theme',
  'txn',
  'budget',
  'attachment',
];

/// manifest.exportKind 的合法取值（§3.2）。
const Set<String> kPayloadExportKinds = <String>{'full', 'range', 'ledger', 'incremental'};

/// 记录行的判别键（§4.1）。
///
/// 判别键在载荷里**必须独占**：任何业务字段都不得叫这个名字。
/// 2026-09-21 裁决的由来 —— §4.1 原文的 account 与 txn 行示例里
/// `type` 出现了两次（判别键 `"account"` 与业务列 `account.type=2`），
/// 而 JSON 对象重复键在 RFC 8259 下行为未定义、实际实现一律取末次。
/// 结果是编码器的 `{'type': stage, ...fields}` 被业务字段反向覆盖：
///
///     {"type":2,"id":"…C1","name":"招行储蓄卡",…}   ← 判别键消失
///
/// 这一行在导入侧既是「未知类型」也是「缺少 type」，两种解析都错。
/// 因此业务列在载荷层的键改名为 `accountType` / `txnType`
/// （只改载荷键，DB 列名不动），并由 [PfbPayloadEncoder.encode] 拒绝
/// 任何携带判别键的记录字段 —— 静默覆盖是这次事故的根因，
/// 换个写法悄悄复现比直接报错危险得多。
const String kPayloadDiscriminatorKey = 'type';

/// 载荷格式版本（与容器 formatVersion 独立演进，§3.2）。
const int kPayloadVersion = 1;

/// 编码结果。
final class PayloadBuildResult {
  const PayloadBuildResult({
    required this.ndjsonBytes,
    required this.recordCount,
    required this.contentHashHex,
  });

  /// 完整载荷：manifest 行 → 记录行 → end 行，每行 `\n` 结尾的 UTF-8。
  final Uint8List ndjsonBytes;

  /// 记录行条数（不含 manifest 与 end）。
  final int recordCount;

  /// contentHash 的十六进制（小写，不带 `sha256:` 前缀）。
  final String contentHashHex;
}

/// NDJSON 载荷编码器。
abstract final class PfbPayloadEncoder {
  /// 把 [stages]（键为 [kPayloadStageOrder] 中的阶段名）编成载荷。
  ///
  /// [manifest] 必须含 `type: "manifest"`，且**不得**已含 `contentHash`；
  /// 缺 `exportKind` 或取值非法、或 scope 专属字段缺失 → `PFC_E_VALIDATION`。
  /// [generatedAtMilliseconds] 是 end 行的 `generatedAt`（显式传入，编码器
  /// 不读时钟 —— 载荷必须可复现）。
  static PayloadBuildResult encode({
    required Map<String, Object?> manifest,
    required Map<String, List<Map<String, Object?>>> stages,
    required int generatedAtMilliseconds,
  }) {
    _validateManifest(manifest);
    for (final stage in stages.keys) {
      if (!kPayloadStageOrder.contains(stage)) {
        throw DomainError.validation(detail: '未知的载荷阶段：$stage');
      }
    }

    // 1. 先编记录行并累计 contentHash（manifest 要携带它，只能后置）。
    //    记录行整体缓冲后一次求摘要 —— M1 载荷本就整体在内存中；
    //    流式分块化（大文件恒定内存）是 M3 接文件网关时的事，届时换
    //    Digest 的分块接口，规则不变。
    final records = BytesBuilder();
    var recordCount = 0;
    for (final stage in kPayloadStageOrder) {
      final rows = stages[stage] ?? const <Map<String, Object?>>[];
      for (final fields in rows) {
        // 判别键必须独占（见 [kPayloadDiscriminatorKey]）。这里刻意用
        // 「拒绝」而不是「覆盖后继续」：`{...fields}` 写在后面会把判别键
        // 抹掉，产出一个**看起来正常、导入侧认不出类型**的载荷 ——
        // 让它在编码期就失败，比让它在半年后的一次真实导入里暴露便宜得多。
        if (fields.containsKey(kPayloadDiscriminatorKey)) {
          throw DomainError.validation(
            detail:
                '阶段 $stage 的记录字段不得包含判别键 "$kPayloadDiscriminatorKey"：'
                '它会覆盖行类型。业务列请在载荷层改名（如 accountType / txnType）',
            userMessage: '导出数据内部不一致，已中止。',
          );
        }
        records.add(_encodeLine(<String, Object?>{kPayloadDiscriminatorKey: stage, ...fields}));
        recordCount++;
      }
    }
    final recordBytes = records.toBytes();
    final contentHashHex = Sha256.instance.hashHex(recordBytes);

    // 2. manifest 行（补 payloadVersion / contentHash）。
    final manifestLine = _encodeLine(<String, Object?>{
      ...manifest,
      'payloadVersion': kPayloadVersion,
      'contentHash': 'sha256:$contentHashHex',
    });

    // 3. end 行。
    final endLine = _encodeLine(<String, Object?>{
      kPayloadDiscriminatorKey: 'end',
      'recordCount': recordCount,
      'contentHash': 'sha256:$contentHashHex',
      'generatedAt': generatedAtMilliseconds,
    });

    final out =
        BytesBuilder()
          ..add(manifestLine)
          ..add(recordBytes)
          ..add(endLine);
    return PayloadBuildResult(
      ndjsonBytes: out.toBytes(),
      recordCount: recordCount,
      contentHashHex: contentHashHex,
    );
  }

  static void _validateManifest(Map<String, Object?> manifest) {
    if (manifest[kPayloadDiscriminatorKey] != 'manifest') {
      throw DomainError.validation(detail: 'manifest 行的 type 必须是 manifest');
    }
    if (manifest.containsKey('contentHash')) {
      throw DomainError.validation(
        detail: 'manifest 不得自带 contentHash（由编码器注入，防止伪造）',
        userMessage: '导出数据内部不一致，已中止。',
      );
    }
    final kind = manifest['exportKind'];
    if (kind is! String || !kPayloadExportKinds.contains(kind)) {
      throw DomainError.validation(detail: 'manifest.exportKind 非法：$kind');
    }
    // scope 专属字段的存在性在编码期就拦（§4.2 第 0 步「前置校验」的载荷侧）。
    switch (kind) {
      case 'range':
        if (manifest['range'] is! Map<String, Object?>) {
          throw DomainError.validation(detail: 'exportKind=range 必须携带 range');
        }
      case 'ledger':
        final ids = manifest['ledgerIds'];
        if (ids is! List<Object?> || ids.isEmpty) {
          throw DomainError.validation(detail: 'exportKind=ledger 必须携带非空 ledgerIds');
        }
      case 'incremental':
        if (manifest['changeLogRange'] is! Map<String, Object?>) {
          throw DomainError.validation(detail: 'exportKind=incremental 必须携带 changeLogRange');
        }
        if (manifest['sinceExportAt'] is! int) {
          throw DomainError.validation(detail: 'exportKind=incremental 必须携带 sinceExportAt');
        }
    }
  }

  static Uint8List _encodeLine(Map<String, Object?> line) {
    // 先白名单校验再序列化：jsonEncode 对不可序列化值抛的是
    // JsonUnsupportedObjectError（Error 子类，禁止捕获）—— 与其压制 lint，
    // 不如把「能不能编码」变成显式的前置校验，报 PFC_E_VALIDATION。
    _ensureJsonable(line, 0);
    return Uint8List.fromList(utf8.encode('${jsonEncode(line)}\n'));
  }

  static void _ensureJsonable(Object? value, int depth) {
    if (value == null || value is String || value is bool || value is num) {
      return;
    }
    if (depth >= 32) {
      throw DomainError.validation(detail: '载荷嵌套深度超过 32', userMessage: '导出数据内部不一致，已中止。');
    }
    if (value is Map<Object?, Object?>) {
      for (final child in value.values) {
        _ensureJsonable(child, depth + 1);
      }
      return;
    }
    if (value is List<Object?>) {
      for (final child in value) {
        _ensureJsonable(child, depth + 1);
      }
      return;
    }
    throw DomainError.validation(
      detail: '载荷含不可 JSON 序列化的值：${value.runtimeType}',
      userMessage: '导出数据内部不一致，已中止。',
    );
  }
}

File: tools/golden_vectors_gen/export_payload.py
#!/usr/bin/env python3
"""导出载荷（NDJSON）黄金向量的生成器 —— 独立参考实现。

## 为什么需要独立实现

`export_payload` 套件锁的是**载荷 NDJSON 的字节**：行序、manifest 的
contentHash 注入位置、end 行字段。期望值若由 Dart 自己算，就是「实现算出
什么就接受什么」。本脚本用 Python 标准库（json / hashlib）独立编出载荷，
与 Dart 侧 `pf_io/src/export_payload.dart` 是两套实现。

JSON 序列化的字节确定性：Dart `jsonEncode` 与 Python
`json.dumps(separators=(",",":"), ensure_ascii=False)` 对「紧凑分隔符、
非 ASCII 原样输出、控制字符转义」的规则一致；测试语料避开两类实现
有分歧的码点（U+2028/2029、 lone surrogate），键序由构造顺序钉死。

## 覆盖范围

- export_payload.full：8 个阶段各 1 条记录（行序 = 父实体先行）+
  manifest 注入 payloadVersion / contentHash + end 行
- export_payload.incremental：增量 manifest（changeLogRange / sinceExportAt）
  只含 txn 阶段
- export_payload.empty：零记录载荷（contentHash = SHA256(空)）

用法：

    python export_payload.py            # 打印 JSON 到标准输出
    python export_payload.py --write    # 写入 test_vectors/v1/
    python export_payload.py --check    # 只做核对，不输出向量
"""

from __future__ import annotations

import argparse
import hashlib
import json
import sys
from pathlib import Path

VECTOR_DIR = Path(__file__).resolve().parents[2] / "test_vectors" / "v1"

STAGE_ORDER = ["ledger", "account", "category", "tag", "theme", "txn", "budget", "attachment"]
PAYLOAD_VERSION = 1

# 记录行的判别键（§4.1）。在载荷里独占：业务列 `account.type` / `txn.type`
# 在载荷层改名为 accountType / txnType（2026-09-21 裁决）。
DISCRIMINATOR = "type"


def dumps_line(line: dict) -> bytes:
    return (json.dumps(line, separators=(",", ":"), ensure_ascii=False) + "\n").encode("utf-8")


def encode(manifest: dict, stages: dict[str, list[dict]], generated_at: int) -> dict:
    """独立实现 §3.2 编码：manifest → 记录行 → end。"""
    assert manifest[DISCRIMINATOR] == "manifest"
    assert "contentHash" not in manifest and "payloadVersion" not in manifest

    record_bytes = bytearray()
    record_count = 0
    for stage in STAGE_ORDER:  # 行序由 STAGE_ORDER 决定，与传入键序无关
        for fields in stages.get(stage, []):
            # 判别键独占（2026-09-21 裁决）。`{"type": stage, **fields}` 里
            # fields 写在后面，一旦它自带 "type" 就会把判别键**静默抹掉** ——
            # 这正是 account / txn 曾经在真实导出里失型的原因。这里断言拦住，
            # 业务列在载荷层的键是 accountType / txnType。
            assert DISCRIMINATOR not in fields, (
                f"阶段 {stage} 的记录字段含判别键 {DISCRIMINATOR!r}："
                "业务列请在载荷层改名（accountType / txnType）"
            )
            record_bytes += dumps_line({DISCRIMINATOR: stage, **fields})
            record_count += 1
    content_hash = hashlib.sha256(bytes(record_bytes)).hexdigest()

    out = bytearray()
    out += dumps_line({**manifest, "payloadVersion": PAYLOAD_VERSION, "contentHash": f"sha256:{content_hash}"})
    out += record_bytes
    out += dumps_line({
        DISCRIMINATOR: "end",
        "recordCount": record_count,
        "contentHash": f"sha256:{content_hash}",
        "generatedAt": generated_at,
    })

    # 自检：解析回去，首行 manifest、末行 end；记录行重序列化后哈希一致
    # （contentHash 只覆盖记录行 —— manifest 携带它、end 引用它，均不在覆盖内）。
    lines = [json.loads(x) for x in bytes(out).decode("utf-8").splitlines()]
    assert lines[0][DISCRIMINATOR] == "manifest" and lines[-1][DISCRIMINATOR] == "end"
    # 每一条记录行的判别键都必须是已知阶段名。
    # 这是 account/txn 撞名事故的回归闸：业务字段一旦覆盖判别键，
    # 该行的 type 会变成整数 1..5，这一句立刻失败。
    assert all(x[DISCRIMINATOR] in STAGE_ORDER for x in lines[1:-1]), (
        "记录行的判别键被覆盖了："
        f"{[x[DISCRIMINATOR] for x in lines[1:-1] if x[DISCRIMINATOR] not in STAGE_ORDER]}"
    )
    assert lines[-1]["recordCount"] == record_count
    record_re = b"".join(dumps_line(x) for x in lines[1:-1])
    assert record_re == bytes(record_bytes)
    assert hashlib.sha256(record_re).hexdigest() == content_hash
    return {
        "ndjsonHex": bytes(out).hex(),
        "recordCount": record_count,
        "contentHashHex": content_hash,
    }


def case(case_id: str, title: str, manifest: dict, stages: dict, generated_at: int,
         notes: str, tags: list[str]) -> dict:
    return {
        "id": case_id,
        "kind": "export.payload.ndjson",
        "title": title,
        "milestone": "M1",
        "input": {"manifest": manifest, "stages": stages, "generatedAt": generated_at},
        "expect": {"ok": True, "value": encode(manifest, stages, generated_at)},
        "notes": notes,
        "tags": tags,
    }


def build_suite() -> dict:
    manifest_full = {
        "type": "manifest",
        "appVersion": "1.0.0",
        "deviceId": "01J8Z9K2M4P6Q8R0T2V4X6Z8B1",
        "deviceName": "向量生成器",
        "platform": "android",
        "exportedAt": 1757922436789,
        "exportKind": "full",
        "includesAttachments": False,
        "counts": {"ledger": 1, "account": 1, "category": 1, "txn": 1, "budget": 1, "tag": 1, "theme": 1, "attachment": 1},
    }
    stages_full = {
        "ledger": [{"id": "01J8TESTLEDGER0000000000001", "name": "日常", "code": "MAIN01", "currency": "CNY", "isDefault": 1, "sortOrder": 0, "createdAt": 1757922436789, "updatedAt": 1757922436789, "deletedAt": None, "deviceId": "01J8Z9K2M4P6Q8R0T2V4X6Z8B1", "rev": 1}],
        "account": [{"id": "01J8TESTACCNT00000000000001", "ledgerId": "01J8TESTLEDGER0000000000001", "name": "招行储蓄卡", "accountType": 2, "currency": "CNY", "openingBalanceMinor": 0, "cachedBalanceMinor": 1284500, "balanceAsOf": 1757900000000, "creditLimitMinor": None, "statementDay": None, "dueDay": None, "repayAccountId": None, "icon": "card", "color": "#3B82F6", "isArchived": 0, "sortOrder": 1, "note": None, "createdAt": 1, "updatedAt": 1, "deletedAt": None, "deviceId": "01J8Z9K2M4P6Q8R0T2V4X6Z8B1", "rev": 1}],
        "category": [{"id": "01J8TESTCATEG0000000000001", "ledgerId": "01J8TESTLEDGER0000000000001", "parentId": None, "kind": 1, "name": "餐饮", "icon": "restaurant", "color": "#F97316", "isSystem": 1, "isHidden": 0, "sortOrder": 10, "createdAt": 1, "updatedAt": 1, "deletedAt": None, "deviceId": "01J8Z9K2M4P6Q8R0T2V4X6Z8B1", "rev": 1}],
        "tag": [{"id": "01J8TESTTAG000000000000001", "ledgerId": "01J8TESTLEDGER0000000000001", "name": "出差", "color": "#8B5CF6", "createdAt": 1, "updatedAt": 1, "deletedAt": None, "deviceId": "01J8Z9K2M4P6Q8R0T2V4X6Z8B1", "rev": 1}],
        "theme": [{"id": "01J8TESTTHEME0000000000001", "name": "我的配色", "specJson": {"primary": "#3B82F6"}, "isActive": 1, "createdAt": 1, "updatedAt": 1, "deletedAt": None, "deviceId": "01J8Z9K2M4P6Q8R0T2V4X6Z8B1", "rev": 1}],
        "txn": [{"id": "01J8TESTTXN00000000000001", "ledgerId": "01J8TESTLEDGER0000000000001", "txnType": 1, "amountMinor": 3800, "currency": "CNY", "occurredAt": 1757900000000, "dayKey": "2026-09-15", "monthKey": "2026-09", "tzOffsetMin": 480, "accountId": "01J8TESTACCNT00000000000001", "toAccountId": None, "categoryId": "01J8TESTCATEG0000000000001", "merchant": "星巴克", "note": "拿铁", "tags": ["01J8TESTTAG000000000000001"], "feeMinor": 0, "isReimbursable": 0, "excludedFromStats": 0, "createdAt": 1, "updatedAt": 1, "deletedAt": None, "deviceId": "01J8Z9K2M4P6Q8R0T2V4X6Z8B1", "originDeviceId": "01J8Z9K2M4P6Q8R0T2V4X6Z8B1", "rev": 1}],
        "budget": [{"id": "01J8TESTBUDGT0000000000001", "ledgerId": "01J8TESTLEDGER0000000000001", "periodType": 1, "periodKey": "2026-09", "scope": 2, "categoryId": "01J8TESTCATEG0000000000001", "amountMinor": 150000, "currency": "CNY", "rollover": 0, "alertBp": 8000, "createdAt": 1, "updatedAt": 1, "deletedAt": None, "deviceId": "01J8Z9K2M4P6Q8R0T2V4X6Z8B1", "rev": 1}],
        "attachment": [],
    }
    manifest_incremental = {
        "type": "manifest",
        "appVersion": "1.0.0",
        "deviceId": "01J8Z9K2M4P6Q8R0T2V4X6Z8B1",
        "deviceName": "向量生成器",
        "platform": "ios",
        "exportedAt": 1757922500000,
        "exportKind": "incremental",
        "sinceExportAt": 1757000000000,
        "changeLogRange": {"minSeq": 120, "maxSeq": 456},
        "includesAttachments": False,
        "counts": {"txn": 1},
    }
    stages_incremental = {
        "txn": [{"id": "01J8TESTTXN00000000000002", "ledgerId": "01J8TESTLEDGER0000000000001", "txnType": 2, "amountMinor": 9900, "currency": "CNY", "occurredAt": 1757920000000, "dayKey": "2026-09-15", "monthKey": "2026-09", "tzOffsetMin": 480, "accountId": "01J8TESTACCNT00000000000001", "toAccountId": None, "categoryId": None, "merchant": None, "note": None, "tags": [], "feeMinor": 0, "isReimbursable": 0, "excludedFromStats": 0, "createdAt": 1757920000000, "updatedAt": 1757920000000, "deletedAt": None, "deviceId": "01J8Z9K2M4P6Q8R0T2V4X6Z8B1", "originDeviceId": "01J8Z9K2M4P6Q8R0T2V4X6Z8B1", "rev": 3}],
    }

    return {
        "schemaVersion": 1,
        "suite": "export_payload",
        "title": "导出载荷 NDJSON（§3.2 / §4.2 3.3–3.5）",
        "description": "行序（父实体先行）、manifest 的 payloadVersion / contentHash 注入、"
                       "end 行字段全部被 Python 独立实现的字节级快照锁死；"
                       "contentHash 覆盖全部非 manifest 行，是导入端载荷完整性的裁决依据。",
        "cases": [
            case(
                "export.payload.ndjson.full",
                "full 导出：8 阶段行序 + manifest 注入 + end 行",
                manifest_full,
                stages_full,
                1757922437000,
                "行序是导入正确性的前提（txn 引用的 account 必须先导入）；contentHash "
                "把全部非 manifest 行钉进一个摘要，行序或字节漂移都会被导入端拒绝。",
                ["format", "security"],
            ),
            case(
                "export.payload.ndjson.incremental",
                "增量导出：scope 专属 manifest 字段 + 单阶段",
                manifest_incremental,
                stages_incremental,
                1757922500000,
                "增量导出携带 changeLogRange / sinceExportAt，三设备场景下增量链"
                "依赖这些字段与原始 deviceId —— 缺失或漂移会让后续增量永久漏记录。",
                ["format", "security"],
            ),
            case(
                "export.payload.ndjson.empty",
                "零记录载荷：contentHash = SHA256(空串)",
                {**manifest_full, "exportedAt": 1757922436790, "counts": {}},
                {},
                1757922437001,
                "空载荷是合法输入（空账本导出），其 contentHash 必须等于空串摘要 "
                "e3b0c442…—— 若实现走「无记录就不算哈希」的捷径，这里立刻变红。",
                ["format"],
            ),
        ],
    }


def main() -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--write", action="store_true", help="写入 test_vectors/v1/")
    parser.add_argument("--check", action="store_true", help="只核对已写入的文件")
    args = parser.parse_args()
    if args.write and args.check:
        print("--write 与 --check 互斥", file=sys.stderr)
        return 2

    suite = build_suite()
    if args.write:
        path = VECTOR_DIR / "export_payload.json"
        # 显式写 LF（`newline="\n"`）：`Path.write_text` 的默认行为在 Windows 上把
        # `\n` 翻译成 `\r\n`，而 `.gitattributes` 把 `*.json` 规范化成 LF 入库 ——
        # 结果是工作区副本 CRLF、索引 LF，`git status` 永远显示「已修改」而
        # `git diff` 是空的。`.gitattributes` 里那段注释点名的就是这个坑。
        path.write_text(
            json.dumps(suite, ensure_ascii=False, indent=2) + "\n",
            encoding="utf-8",
            newline="\n",
        )
        print(f"wrote {path} ({len(suite['cases'])} cases)")
    elif args.check:
        path = VECTOR_DIR / "export_payload.json"
        if not path.exists():
            print(f"missing {path}", file=sys.stderr)
            return 1
        existing = json.loads(path.read_text(encoding="utf-8"))
        if existing != suite:
            print(f"MISMATCH {path.name}", file=sys.stderr)
            return 1
        print(f"checked {path.name} ({len(suite['cases'])} cases)")
    else:
        print(json.dumps(suite, ensure_ascii=False, indent=2))
    print(f"total {len(suite['cases'])} cases")
    return 0


if __name__ == "__main__":
    sys.exit(main())
